### PR TITLE
Make oodeel library agnostic

### PR DESCRIPTION
**Pull request by paulnovello**
library = tensorflow or pytorch

Tasks:

- [x]  Remove library-specific syntax from OODmodel
- [x]  Implement other baselines with pytorch
- [x] Make sure correct tensors are passed as input to oodmodel
- [x] Manage detach
- [x] replace elem by item in signature names
- [x] Include tf_utils in tf operator or data_handler
- [x] Use pytorch synthax in operators
- [x]  Test everything on new notebooks
- [x] Make sure one can use the code with only pytorch or tensorflow installed (create additional test envs)
- [ ] double check types (left to next PR - this one is too big)

### CODE REVIEW

File: tests/tests_tensorflow/methods/test_tf_dknn.py [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+import pytest
+
+from oodeel.methods import DKNN
+from tests.tests_tensorflow import generate_data
+from tests.tests_tensorflow import generate_data_tf
+from tests.tests_tensorflow import generate_model
+
+
+@pytest.mark.parametrize(
+    ("backend", "input_shape"),
```
Comment (y-prudent) [inline tests/tests_tensorflow/methods/test_tf_dknn.py:32]: The backend arg is useless since only tf models and datasets are considered.

**Review by y-prudent — COMMENTED** [review]
Great job! Main remarks concern:
- Docstrings to update
- Some unessential code?
- TODO for VIM: use TF/torch operators instead of numpy to speed up computation
- For the tests, the main changes will consist in removing all the `backend` unnecessary arguments and conditions related to this arg

File: oodeel/models/training_funs_torch/torch_models.py [hunk 2]
```diff
@@ -155,31 +170,32 @@ def _augment_fn(images, labels):
...
-        save_dir (_type_): _description_
-        device (_type_): _description_
+        model (nn.Module): Model to train.
+        train_data (DataLoader): Train dataloader.
+        epochs (int): Number of training epochs.
+        optimizer (torch.optim.Optimizer): Optimizer.
+        lr_scheduler (torch.optim.lr_scheduler._LRScheduler): Learning rate scheduler.
+            Defaults to None.
+        criterion (torch.nn.modules.loss._Loss): Criterion for loss.
+        device (torch.device): On which device to train (CUDA or CPU).
```
Comment (cofri) [inline oodeel/models/training_funs_torch/torch_models.py:193]: I don't know if it is important for documentation but criterion and device docstrings are after lr_scheduler, whereas it's the opposite in the function's signature.

File: oodeel/models/torch_feature_extractor.py [hunk 3]
```diff
@@ -129,7 +133,10 @@ def prepare_extractor(self):
...
                 raise NotImplementedError
 
-    def predict_tensor(self, x: torch.Tensor) -> List[torch.Tensor]:
+    @sanitize_input
+    def predict_tensor(
+        self, x: torch.Tensor, detach: bool = True
+    ) -> List[torch.Tensor]:
         """Get the projection of tensor in the feature space of self.model
 
         Args:
```
Comment (cofri) [inline oodeel/models/torch_feature_extractor.py:142]: Missing docstring for argument `detach`

File: oodeel/utils/tf_operator.py [hunk 4]
```diff
@@ -57,6 +73,33 @@ def sign(tensor: Union[tf.Tensor, np.ndarray]) -> tf.Tensor:
         return tf.sign(tensor)
 
     @staticmethod
+    def CrossEntropyLoss(reduction: str = "mean"):
```
Comment (cofri) [inline oodeel/utils/tf_operator.py:76]: `reduction="mean"` is not a supported argument for Keras losses. Maybe do you mean `sum_over_batch_size`? Or even `auto`?

File: README.md [hunk 5]
```diff
@@ -12,7 +12,7 @@
 <!-- Badge section -->
 <div align="center">
     <a href="#">
-        <img src="https://img.shields.io/badge/Python-3.7, 3.8, 3.9, 3.10-efefef">
+        <img src="https://img.shields.io/badge/Python-3.8, 3.9, 3.10-efefef">
```
Comment (cofri) [inline README.md:15]: Same as `index.md`, we can instead use https://img.shields.io/badge/Python-3.8+-efefef for more supported Python versions

**Review by cofri — CHANGES_REQUESTED** [review]
Nice PR! Agnostic is all you need
I suggested some changes (see comments). I haven't reviewed tests yet. I will review them this afternoon.

File: tests/tests_tensorflow/datasets/test_tf_ooddataset.py [hunk 6]
```diff
@@ -98,23 +74,17 @@ def test_instanciate_ood_dataset(backend, as_supervised, expected_output):
     [
         ("tensorflow", True, [200, 2, 3, 0.5]),
```
Comment (cofri) [inline tests/tests_tensorflow/datasets/test_tf_ooddataset.py:75]: If only "tensorflow" backend is used in this file, can we remove it from the pytest parametrization?

File: tests/tests_torch/methods/test_torch_mls.py [hunk 7]
```diff
@@ -20,30 +20,39 @@
...
+from tests.tests_torch import generate_data
+from tests.tests_torch import generate_data_torch
+
+
+@pytest.mark.parametrize(
+    ("backend", "input_shape"),
+    [("torch", (3, 32, 32))],
+    ids=["[torch] test DKNN"],
+)
+def test_dknn(backend, input_shape):
```
Comment (cofri) [inline tests/tests_torch/methods/test_torch_mls.py:37]: There are several references to DKNN here that should be replaced with MLS instead: function's name, docstring, OODModel instance's name

File: tests/tests_torch/methods/test_torch_mls.py [hunk 8]
```diff
@@ -20,30 +20,39 @@
...
-
-def test_isood():
-    input_shape = (32, 32, 3)
+from tests.tests_torch import ComplexNet
+from tests.tests_torch import generate_data
+from tests.tests_torch import generate_data_torch
+
+
+@pytest.mark.parametrize(
+    ("backend", "input_shape"),
```
Comment (cofri) [inline tests/tests_torch/methods/test_torch_mls.py:33]: Here also, `backend` parameter could be removed

File: tests/tests_torch/methods/test_torch_odin.py [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+from tests.tests_torch import generate_data
+from tests.tests_torch import generate_data_torch
+
+
+@pytest.mark.parametrize(
+    ("backend", "input_shape"),
+    [("torch", (3, 32, 32))],
+    ids=["[torch] test DKNN"],
+)
+def test_dknn(backend, input_shape):
```
Comment (cofri) [inline tests/tests_torch/methods/test_torch_odin.py:37]: Same here, multiple references to DKNN instead of ODIN

File: tests/tests_torch/methods/test_torch_vim.py [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+from tests.tests_torch import generate_data
+from tests.tests_torch import generate_data_torch
+
+
+@pytest.mark.parametrize(
+    ("backend", "input_shape"),
+    [("torch", (3, 32, 32))],
+    ids=["[torch] test DKNN"],
+)
+def test_dknn(backend, input_shape):
```
Comment (cofri) [inline tests/tests_torch/methods/test_torch_vim.py:37]: Same here, multiple references to DKNN instead of VIM

**Review by cofri — CHANGES_REQUESTED** [review]
Second part of the review for "tests" folder

File: oodeel/methods/base.py [hunk 11]
```diff
@@ -106,17 +105,21 @@ def _load_feature_extractor(
                 keras models saved as pb files e.g. with model.save()
         """
         if is_from(model, "keras"):
-            from ..utils import TFOperator
             from ..models.keras_feature_extractor import KerasFeatureExtractor
+            from ..datasets.tf_data_handler import TFDataHandler
+            from ..utils import TFOperator
 
+            self.dh = TFDataHandler()
```
[the remark this answers is not in the document]
Comment (y-prudent) [reply oodeel/methods/base.py:108]: @paulnovello

File: oodeel/models/training_funs_torch/train_cifar10.py
[the remark this answers is not in the document]
Comment (paulnovello) [reply oodeel/models/training_funs_torch/train_cifar10.py:1]: Not really, I am deleting it

File: oodeel/methods/vim.py
[the remark this answers is not in the document]
Comment (paulnovello) [reply oodeel/methods/vim.py:1]: Agree to leave it for a future PR

File: oodeel/datasets/ooddataset.py
[the remark this answers is not in the document]
Comment (paulnovello) [reply oodeel/datasets/ooddataset.py:1]: For now I suggest adding a call to `get_dataset` in the `prepare` function to avoid code repetition. This function is used when using `training_func` with is `prepared=False`. I agree that it is discussable, but I leave it for a future PR.
For `self._data_handler.dict_to_tuple`, could you add it to torch_data_handler ?

**Review by cofri — APPROVED** [review]

**Review by y-prudent — APPROVED** [review]
LGTM!
